File: src/frontend/financeiro/lib/app/data/model/fin_cheque_recebido_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:financeiro/app/infra/infra_imports.dart';
import 'package:financeiro/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class FinChequeRecebidoModel {
	int? id;
	int? idCliente;
	String? cpf;
	String? cnpj;
	String? nome;
	String? codigoBanco;
	String? codigoAgencia;
	String? conta;
	int? numero;
	DateTime? dataEmissao;
	DateTime? bomPara;
	DateTime? dataCompensacao;
	double? valor;
	DateTime? custodiaData;
	double? custodiaTarifa;
	double? custodiaComissao;
	DateTime? descontoData;
	double? descontoTarifa;
	double? descontoComissao;
	double? valorRecebido;
	ViewPessoaClienteModel? viewPessoaClienteModel;

	FinChequeRecebidoModel({
		this.id,
		this.idCliente,
		this.cpf,
		this.cnpj,
		this.nome,
		this.codigoBanco,
		this.codigoAgencia,
		this.conta,
		this.numero,
		this.dataEmissao,
		this.bomPara,
		this.dataCompensacao,
		this.valor,
		this.custodiaData,
		this.custodiaTarifa,
		this.custodiaComissao,
		this.descontoData,
		this.descontoTarifa,
		this.descontoComissao,
		this.valorRecebido,
		this.viewPessoaClienteModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'cpf',
		'cnpj',
		'nome',
		'codigo_banco',
		'codigo_agencia',
		'conta',
		'numero',
		'data_emissao',
		'bom_para',
		'data_compensacao',
		'valor',
		'custodia_data',
		'custodia_tarifa',
		'custodia_comissao',
		'desconto_data',
		'desconto_tarifa',
		'desconto_comissao',
		'valor_recebido',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Cpf',
		'Cnpj',
		'Nome',
		'Codigo Banco',
		'Codigo Agencia',
		'Conta',
		'Numero',
		'Data Emissao',
		'Bom Para',
		'Data Compensacao',
		'Valor',
		'Custodia Data',
		'Custodia Tarifa',
		'Custodia Comissao',
		'Desconto Data',
		'Desconto Tarifa',
		'Desconto Comissao',
		'Valor Recebido',
	];

	FinChequeRecebidoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCliente = jsonData['idCliente'];
		cpf = jsonData['cpf'];
		cnpj = jsonData['cnpj'];
		nome = jsonData['nome'];
		codigoBanco = jsonData['codigoBanco'];
		codigoAgencia = jsonData['codigoAgencia'];
		conta = jsonData['conta'];
		numero = jsonData['numero'];
		dataEmissao = jsonData['dataEmissao'] != null ? DateTime.tryParse(jsonData['dataEmissao']) : null;
		bomPara = jsonData['bomPara'] != null ? DateTime.tryParse(jsonData['bomPara']) : null;
		dataCompensacao = jsonData['dataCompensacao'] != null ? DateTime.tryParse(jsonData['dataCompensacao']) : null;
		valor = jsonData['valor']?.toDouble();
		custodiaData = jsonData['custodiaData'] != null ? DateTime.tryParse(jsonData['custodiaData']) : null;
		custodiaTarifa = jsonData['custodiaTarifa']?.toDouble();
		custodiaComissao = jsonData['custodiaComissao']?.toDouble();
		descontoData = jsonData['descontoData'] != null ? DateTime.tryParse(jsonData['descontoData']) : null;
		descontoTarifa = jsonData['descontoTarifa']?.toDouble();
		descontoComissao = jsonData['descontoComissao']?.toDouble();
		valorRecebido = jsonData['valorRecebido']?.toDouble();
		viewPessoaClienteModel = jsonData['viewPessoaClienteModel'] == null ? ViewPessoaClienteModel() : ViewPessoaClienteModel.fromJson(jsonData['viewPessoaClienteModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCliente'] = idCliente != 0 ? idCliente : null;
		jsonData['cpf'] = Util.removeMask(cpf);
		jsonData['cnpj'] = Util.removeMask(cnpj);
		jsonData['nome'] = nome;
		jsonData['codigoBanco'] = codigoBanco;
		jsonData['codigoAgencia'] = codigoAgencia;
		jsonData['conta'] = conta;
		jsonData['numero'] = numero;
		jsonData['dataEmissao'] = dataEmissao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataEmissao!) : null;
		jsonData['bomPara'] = bomPara != null ? DateFormat('yyyy-MM-ddT00:00:00').format(bomPara!) : null;
		jsonData['dataCompensacao'] = dataCompensacao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataCompensacao!) : null;
		jsonData['valor'] = valor;
		jsonData['custodiaData'] = custodiaData != null ? DateFormat('yyyy-MM-ddT00:00:00').format(custodiaData!) : null;
		jsonData['custodiaTarifa'] = custodiaTarifa;
		jsonData['custodiaComissao'] = custodiaComissao;
		jsonData['descontoData'] = descontoData != null ? DateFormat('yyyy-MM-ddT00:00:00').format(descontoData!) : null;
		jsonData['descontoTarifa'] = descontoTarifa;
		jsonData['descontoComissao'] = descontoComissao;
		jsonData['valorRecebido'] = valorRecebido;
		jsonData['viewPessoaClienteModel'] = viewPessoaClienteModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCliente = plutoRow.cells['idCliente']?.value;
		cpf = plutoRow.cells['cpf']?.value;
		cnpj = plutoRow.cells['cnpj']?.value;
		nome = plutoRow.cells['nome']?.value;
		codigoBanco = plutoRow.cells['codigoBanco']?.value;
		codigoAgencia = plutoRow.cells['codigoAgencia']?.value;
		conta = plutoRow.cells['conta']?.value;
		numero = plutoRow.cells['numero']?.value;
		dataEmissao = Util.stringToDate(plutoRow.cells['dataEmissao']?.value);
		bomPara = Util.stringToDate(plutoRow.cells['bomPara']?.value);
		dataCompensacao = Util.stringToDate(plutoRow.cells['dataCompensacao']?.value);
		valor = plutoRow.cells['valor']?.value?.toDouble();
		custodiaData = Util.stringToDate(plutoRow.cells['custodiaData']?.value);
		custodiaTarifa = plutoRow.cells['custodiaTarifa']?.value?.toDouble();
		custodiaComissao = plutoRow.cells['custodiaComissao']?.value?.toDouble();
		descontoData = Util.stringToDate(plutoRow.cells['descontoData']?.value);
		descontoTarifa = plutoRow.cells['descontoTarifa']?.value?.toDouble();
		descontoComissao = plutoRow.cells['descontoComissao']?.value?.toDouble();
		valorRecebido = plutoRow.cells['valorRecebido']?.value?.toDouble();
		viewPessoaClienteModel = ViewPessoaClienteModel();
		viewPessoaClienteModel?.nome = plutoRow.cells['viewPessoaClienteModel']?.value;
	}	

	FinChequeRecebidoModel clone() {
		return FinChequeRecebidoModel(
			id: id,
			idCliente: idCliente,
			cpf: cpf,
			cnpj: cnpj,
			nome: nome,
			codigoBanco: codigoBanco,
			codigoAgencia: codigoAgencia,
			conta: conta,
			numero: numero,
			dataEmissao: dataEmissao,
			bomPara: bomPara,
			dataCompensacao: dataCompensacao,
			valor: valor,
			custodiaData: custodiaData,
			custodiaTarifa: custodiaTarifa,
			custodiaComissao: custodiaComissao,
			descontoData: descontoData,
			descontoTarifa: descontoTarifa,
			descontoComissao: descontoComissao,
			valorRecebido: valorRecebido,
		);			
	}

	
}

File: backend/financeiro/src/model/fin_cheque_recebido_model.py
from src import db
from src.model.view_pessoa_cliente_model import ViewPessoaClienteModel


class FinChequeRecebidoModel(db.Model):
    __tablename__ = 'fin_cheque_recebido'

    id = db.Column(db.Integer, primary_key=True)
    cpf = db.Column(db.String(11))
    cnpj = db.Column(db.String(14))
    nome = db.Column(db.String(100))
    codigo_banco = db.Column(db.String(10))
    codigo_agencia = db.Column(db.String(10))
    conta = db.Column(db.String(20))
    numero = db.Column(db.Integer)
    data_emissao = db.Column(db.DateTime)
    bom_para = db.Column(db.DateTime)
    data_compensacao = db.Column(db.DateTime)
    valor = db.Column(db.Float)
    custodia_data = db.Column(db.DateTime)
    custodia_tarifa = db.Column(db.Float)
    custodia_comissao = db.Column(db.Float)
    desconto_data = db.Column(db.DateTime)
    desconto_tarifa = db.Column(db.Float)
    desconto_comissao = db.Column(db.Float)
    valor_recebido = db.Column(db.Float)
    id_cliente = db.Column(db.Integer, db.ForeignKey('view_pessoa_cliente.id'))

    view_pessoa_cliente_model = db.relationship('ViewPessoaClienteModel', foreign_keys=[id_cliente])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_cliente = data.get('idCliente')
        self.cpf = data.get('cpf')
        self.cnpj = data.get('cnpj')
        self.nome = data.get('nome')
        self.codigo_banco = data.get('codigoBanco')
        self.codigo_agencia = data.get('codigoAgencia')
        self.conta = data.get('conta')
        self.numero = data.get('numero')
        self.data_emissao = data.get('dataEmissao')
        self.bom_para = data.get('bomPara')
        self.data_compensacao = data.get('dataCompensacao')
        self.valor = data.get('valor')
        self.custodia_data = data.get('custodiaData')
        self.custodia_tarifa = data.get('custodiaTarifa')
        self.custodia_comissao = data.get('custodiaComissao')
        self.desconto_data = data.get('descontoData')
        self.desconto_tarifa = data.get('descontoTarifa')
        self.desconto_comissao = data.get('descontoComissao')
        self.valor_recebido = data.get('valorRecebido')

    def serialize(self):
        return {
            'id': self.id,
            'idCliente': self.id_cliente,
            'cpf': self.cpf,
            'cnpj': self.cnpj,
            'nome': self.nome,
            'codigoBanco': self.codigo_banco,
            'codigoAgencia': self.codigo_agencia,
            'conta': self.conta,
            'numero': self.numero,
            'dataEmissao': self.data_emissao.isoformat(),
            'bomPara': self.bom_para.isoformat(),
            'dataCompensacao': self.data_compensacao.isoformat(),
            'valor': self.valor,
            'custodiaData': self.custodia_data.isoformat(),
            'custodiaTarifa': self.custodia_tarifa,
            'custodiaComissao': self.custodia_comissao,
            'descontoData': self.desconto_data.isoformat(),
            'descontoTarifa': self.desconto_tarifa,
            'descontoComissao': self.desconto_comissao,
            'valorRecebido': self.valor_recebido,
            'viewPessoaClienteModel': self.view_pessoa_cliente_model.serialize() if self.view_pessoa_cliente_model else None,
        }